examples/vision/barlow_twins.py

```diff
@@ -0,0 +1,1046 @@
+
+It also has the advantage of being much simpler than other methods, and its
+model architecture is symmetric, meaning that both twins in the model do the
+same thing. It is also near state-of-the-art on imagenet, even exceeding methods
+like SIM-clr.
+
+
+One disadvantage of Barlow Twins is that it is heavily dependent on
+augmentation, suffering major performance decreases in accuracy without them.
+
+TL, DR: Barlow twins creates representations that are:
+*   Invariant.
+*   Not redundant, and carry as much info about the dataset.
+
+Also, it is simpler than other methods.
+
+This notebook was able to train a barlow twins model and reach approximately
+64% validation accuracy.
+"""
+
+"""
+![image](https://i.imgur.com/G6LnEPT.png)
+"""
+
+"""
+Taken from the original paper:
+[Barlow Twins: Self-Supervised Learning viaRedundancy Reduction](https://arxiv.org/abs/2103.03230)
+"""
+
+"""
+# High-Level Theory
+"""
+
+"""
+The model takes two versions of the same image(with different augmentations) as
+input. Then it takes a prediction of each of them, creating representations.
+They are then used to make a cross-correlation matrix.
+
+Cross correlation matrix:
+```
+(pred_1.T @ pred_2) / batch_size
+```
+
+The cross-correlation matrix measures the correlation between the output
+neurons in the two representations made by the model predictions of the two
+augmented versions of data. Ideally, a cross-correlation matrix should look
+like an identity matrix if the two images are the same.
+
+When this happens, it means that the representations:
+1.   Are invariant. The diagonal shows the correlation between each
+representation's neurons and its corresponding augmented one. Because the two
+versions come from the same image, the diagonal of the matrix should show that
+there is a strong correlation between them. If the images are different, there
+shouldn't really be a diagonal.
+2.   Do not show signs of redundancy. If the neurons show correlation with a
+non-diagonal neuron, it means that it is not correctly identifying similarities
+between the two augmented images. This means that it is redundant.
+
+Here is a good way of understanding in pseudocode(information from the original
+paper):
+
+```
+c[i][i] = 1
+c[i][j] = 0
+
+where:
+  c is the cross-correlation matrix
+  i is the index of one representation's neuron
+  j is the index of the second representation's neuron
+```
+"""
+
+"""
+# Setup
+"""
+
+"""shell
+!pip install tensorflow-addons
+"""
+
+import os
+
+os.environ["TF_GPU_THREAD_MODE"] = "gpu_private"  # for faster performance
+
+import tensorflow as tf  # framework
+import tensorflow.keras as keras  # for tf.keras
+import tensorflow_addons as tfa  # LAMB optimizer and gaussian_blur_2d function
+import numpy as np  # np.random.random
+import matplotlib.pyplot as plt  # graphs
+import datetime  # tensorboard logs naming
+
+tf.config.optimizer.set_jit(True)  # XLA optimization for faster performance
```

Comment:
All of my experiment until now with `jit = True` wasn't satisfactory. I'm curious to know, have you found it useful generally? I'll do some experiments with it.

Reply:
I believe it had some performance improvement saving a few seconds each epoch but I'll formally check.

Reply:
Saves approx 4 seconds per epoch. So saves 10-15 minutes of training time